### PR TITLE
Refactor the ApplicationManager to use the AddonManager to get applications info

### PR DESCRIPTION
**Pull request by DaazKu**
That way applications can work with both about.php and addon.json.

### CODE REVIEW

**Comment by initvector** [comment]
On hold pending a fix for the build

File: library/Vanilla/Addon.php [hunk 1]
```diff
@@ -144,6 +144,12 @@ private function scanInfo() {
             if (empty($info)) {
                 throw new \Exception("The addon at $subdir has an empty info array.");
             }
+
+            $baseDir = explode('/', ltrim($subdir, '/'))[0];
+            if (in_array($baseDir, ['applications', 'plugins'])) {
```
Comment (tburry) [inline library/Vanilla/Addon.php]: I don't want the new Addon class to have any knowledge of the directory it is in.

**Comment by DaazKu** [comment]
I created an [issue about removing the kludge](https://github.com/vanilla/vanilla/issues/5552) as part of [Application/Plugin unification and enhancement](https://github.com/vanilla/vanilla/issues/5106)

File: library/core/class.applicationmanager.php [hunk 2]
```diff
@@ -48,35 +48,26 @@ public function __construct(AddonManager $addonManager = null) {
...
-                        $Folder = substr($Folder, 0, strpos($Folder, DS));
-                        $ApplicationInfo[$ApplicationName]['Folder'] = $Folder;
-                    }
+            $applications = [];
+
+            $addons = $this->addonManager->lookupAllByType(Addon::TYPE_ADDON);
+            foreach ($addons as $addon) {
+                /* @var Addon $addon */
+                if ($addon->getInfoValue('oldType') !== 'application') {
+                    continue;
```
Comment (tburry) [inline library/core/class.applicationmanager.php]: Right away you have some issues here. You need to make Gdn_ApplicationManager backwards compatible or else some plugin somewhere is going to break. Look at how Gdn_PluginManager does it:

```php
public function availablePlugins() {
    $addons = $this->addonManager->lookupAllByType(Addon::TYPE_ADDON);

    $result = [];
    foreach ($addons as $addon) {
        /* @var Addon $addon */
        if ($addon->getInfoValue('oldType') !== 'plugin') {
            continue;
        }

        $infoArray  = static::calcOldInfoArray($addon);
        $result[$infoArray['Index']] = $infoArray;
    }
    return $result;
}
```

You can do something almost exactly the same and use Gdn_PluginManager::calcOldInfoArray() too.

Comment (DaazKu) [inline library/core/class.applicationmanager.php]: I already do it...
```
$info['Folder'] = $directories[count($directories) - 1];
$info['Index'] = $info['name'];
```

**Comment by DaazKu** [comment]
The application manager is now 100% backward compatible.

**Pull request merged**